python/pyarrow/array.pxi

```diff
@@ -2478,6 +2478,55 @@ cdef class DictionaryArray(Array):
 
         return self._indices
 
+    @staticmethod
+    def from_buffers(DataType type, int64_t length, buffers, dictionary,
+                     int64_t null_count=-1, int64_t offset=0, memory_pool=None):
+        """
+        Construct a DictionaryArray from buffers.
+
+        type : pyarrow.DataType
+        length : int
+            The number of values in the array.
+        buffers : List[Buffer]
+            The buffers backing this array.
+        dictionary : pyarrow.Array, ndarray or pandas.Series
+            The array of values referenced by the indices.
+        null_count : int, default -1
+            The number of null entries in the array. Negative value means that
+            the null count is not known.
+        offset : int, default 0
+            The array's logical offset (in values, not in bytes) from the
+            start of each buffer.
+        memory_pool : MemoryPool, default None
+            For memory allocations, if required, otherwise uses default pool.
+        """
+        cdef:
+            Array _dictionary
+            vector[shared_ptr[CBuffer]] c_buffers
+            shared_ptr[CDataType] c_type
+            shared_ptr[CArrayData] c_data
+            shared_ptr[CArray] c_result
+
+        for buf in buffers:
+            c_buffers.push_back(pyarrow_unwrap_buffer(buf))
+
+        if isinstance(dictionary, Array):
+            _dictionary = dictionary
+        else:
+            _dictionary = array(dictionary, memory_pool=memory_pool)
```

Comment:
Given that this is a "power-user" method anyway, I think it is maybe fine to require an Array here for the dictionary (for the Buffers we also only accept actual buffers, and that would avoid needing to add the `memory_pool` keyword, and avoid having to test this, which isn't the case at the moment)